### PR TITLE
Upgrade to zlib >=0.6.

### PR DESCRIPTION
**Pull request by mf59816**
In order to link to hackage-security, we need to bump zlib version and adjust to api changes.

I am not familiar with cabal coding guidelines and habits, but I appreciate any suggestions.  Also, I have questions:

1. Should I write a test?  (Haven't found one on this.)
2. Is there really a need for zlib compression?  If we only had to do gzip, things would look a lot easier.
3. Should we really trust the gzip magic header?  It's just two bytes, and they may match for some random reason even on unzipped data.
4. Should I throw out the gzip case?  The zlib case should handle both as before.

Thanks,
m.

### CODE REVIEW

File: cabal-install/cabal-install.cabal [hunk 1]
```diff
@@ -142,7 +142,7 @@ executable cabal
         random     >= 1        && < 1.2,
         stm        >= 2.0      && < 3,
         time       >= 1.1      && < 1.6,
-        zlib       >= 0.5.3    && < 0.6
+        zlib       >= 0.6
```
Comment (23Skidoo) [inline cabal-install/cabal-install.cabal]: Please also add an upper bound to comply with PVP.

**Comment by 23Skidoo** [comment]
Having a test would be nice.

**Comment by mf59816** [comment]
I changed the entire thing again after @dcoutts asked me to keep support for <0.6.

I can't see any way of making the change smaller, but that may be due to my limited understanding of the zlib api (old or new).  @dcoutts: if you give me a hint, I'll try harder.

No test case yet (sorry), but I did test both zlib 0.5 and 0.6 locally, and with another small fix both CPP-branches downloaded and unpacked 00-index.tar ok.  Let see what travis thinks of this...

**Comment by mf59816** [comment]
(The commit is in two parts so travis will link both against zlib 0.5 and 0.6.  I hope I got it right the third time...)

**Comment by dcoutts** [comment]
LGTM

**Comment by dcoutts** [comment]
Could do a unit test, it'd look a lot like this:

    maybeDecompressUnitTest :: Bool
    maybeDecompressUnitTest =
        maybeDecompress original == original
     && maybeDecompress compressedZlib == original
     && maybeDecompress compressedGZip == original
      where
        original = BS.pack "original uncompressed input"
        compressedZlib = ZLib.compress original
        compressedGZip = GZip.compress original

**Comment by mf59816** [comment]
Added.  Thanks for the code (-:

(now i am feeling a bit lazy.  sorry...)

**Comment by mf59816** [comment]
erm...  that is not going to work.  have to go, though.  i'll try again tomorrow.

**Comment by edsko** [comment]
This is not correct, I think. We should not return the input stream on _any_ error, but only when there is an error at the start of the input stream; if there are errors halfway the stream we should throw them as usual. See https://github.com/edsko/cabal/blob/using-hackage-security/cabal-install/Distribution/Client/GZipUtils.hs#L27-68 for an alternative implementation.

**Comment by mf59816** [comment]
I made travis test both zlib < 0.6 and >= 0.6 again, and expect them both to go through with @edsko's implementation of `maybeDecompress`.

However: I had to compare equality of `show` of the bytestrings in the test suite.  The bytestrings themselves appear to not be equal.  Not sure how this can not be a bug in bytestring.  Somehow I hope it is something about my local setup, and travis will pass all three commits.

**Comment by mf59816** [comment]
The travis errors in https://github.com/zerobuzz/cabal/commit/6d71b4207874877aa9121fa417de5b8ae4894a4e can be explained by https://github.com/haskell/cabal/commit/94ac1e0204786978cc8079eea57d07400c7655a3.

https://github.com/zerobuzz/cabal/commit/27ea7082e334251add9542c56b8bdcc12430d688 surprises me: the failing test case "decompress zlib" does not even exist in this commit; it should have been replaced by "decompress zlib (with show)".

The commits in the travis logs don't match the ones from this pr.  How does travis get the commit to test?

**Comment by snoyberg** [comment]
This is currently blocking Stackage from upgrading to zlib 0.6: https://github.com/fpco/stackage/issues/537

**Comment by 23Skidoo** [comment]
Travis build fails, so I can't merge this yet.

**Comment by mf59816** [comment]
sorry..  i've been meaning to look into the bizarre string unequality issue, but to be realistic it won't happen this week.

**Comment by phadej** [comment]
There seems to be something wrong with `ByteString` `Eq`:

```haskell
λ > import Codec.Compression.GZip          as GZip
λ > import Codec.Compression.Zlib          as Zlib
λ > import Distribution.Client.GZipUtils           (maybeDecompress)
λ > let original = BSChar8.pack "foobar"
λ > let compressedZlib = Zlib.compress original
λ > let compressedGZip = GZip.compress original
λ > maybeDecompress compressedZlib        == original
False
λ > maybeDecompress compressedGZip        == original
False
λ > maybeDecompress compressedZlib
"foobar"
λ > maybeDecompress compressedGZip
"foobar"
λ > BS.unpack (maybeDecompress compressedZlib)        == BS.unpack original
True
λ > BS.unpack (maybeDecompress compressedGZip)        == BS.unpack original
λ > BS.unpack (maybeDecompress compressedGZip )
[102,111,111,98,97,114]
λ > BS.unpack original
[102,111,111,98,97,114]
```

That because lazy bytestring is [compared by chunks](https://github.com/haskell/bytestring/blob/c2ddcf96cdc6bf83206457a781801f6ff45a1aa6/Data/ByteString/Lazy/Internal.hs#L225)
```
λ > toChunks (maybeDecompress compressedGZip )
["foobar",""]
λ > toChunks original 
["foobar"]
````

**Comment by phadej** [comment]
I corrected a test case in https://github.com/phadej/cabal/tree/upgrade-zlib (it's also rebased to current master), https://travis-ci.org/phadej/cabal/builds/68143393 is a travis job building it right now.

**Comment by 23Skidoo** [comment]
```
λ > maybeDecompress compressedGZip        == original
False
λ > toChunks (maybeDecompress compressedGZip )
["foobar",""]
λ > toChunks original 
["foobar"]
````

Wow, that's news to me! @dcoutts, shouldn't the `bytestring` library be fixed? And perhaps `zlib` shouldn't produce zero-length chunks in the first place.

**Comment by edsko** [comment]
This sounds like a bug in `zlib` actually -- lazy bytestrings should never have zero-length chunks.

**Comment by edsko** [comment]
Or possibly a bug in this patch, let me think about that..

**Comment by edsko** [comment]
No, the only chunks we create here use bytestrings returned by `zlib` as part of the `DecompressStream`; I don't think the bug is in the patch. (But @dcoutts did mention that there was a problem with zlib 0.6; maybe this is related?)

**Comment by phadej** [comment]
@23Skidoo IMHO we should change that in `bytestring`. We can observe the difference thru `toChunks`, which isn't a *debugging* function like [Data.Map.showTree](http://hackage.haskell.org/package/containers-0.5.6.3/docs/Data-Map-Lazy.html#v:showTree)

**Comment by edsko** [comment]
No, I don't think so. This is an invariant violation.

**Comment by phadej** [comment]
@edsko sorry, I meant we shouldn't change lazy bytestring comparison. We could have `["foo", "bar"]` and `["foobar"]` bytestrings.

Another question is what to do with empty chunks. Maybe `Builder` should omit them? Don't know if `zlib` uses it though.

**Comment by phadej** [comment]
Found a reason. in latest zlib:

```haskell
decompress   format params = foldDecompressStreamWithInput
                               L.Chunk (const L.Empty) throw
                               (decompressStreamST format params)
```
`const L.empty` is `end`
```haskell
-- in foldDecompressStreamWithInput:
    fold (DecompressStreamEnd inchunk) inchunks =
      return $ end (L.fromChunks (inchunk:inchunks))
```



But we do:
```haskell
go _  (DecompressStreamEnd       bs  ) = return $ Chunk bs Empty
```

This is easy to fix.

**Comment by phadej** [comment]
I made a small fix to `maybeDecompress`: https://github.com/phadej/cabal/commit/9037abd4ea15a6002151e1659af4ff8087534f58

cabal build: https://travis-ci.org/phadej/cabal/builds/68207545 (I'm afraid ghc 7.4 / bytestring 0.9 `with show` test will fail)

**Comment by 23Skidoo** [comment]
Closing in favour of #2693.

**Pull request closed without merging**